File: Project1/Python/MergeSort.py
"""
Name: Alex Bryant
ID: 2
Professor: Dr. Dai
Course: CS416 - Software Engineering
Date: Feb 19, 2026

File: MergeSort.py
"""

def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]
        merge_sort(left)
        merge_sort(right)
        i = j = k = 0
        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1
            k += 1
        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1
        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1
    return arr

arr = [64, 34, 25, 12, 22, 11, 90]
print("Sample Input: ", arr)
merge_sort(arr)
print("Sample Output: ", arr)

File: Project1/Go/MergeSort.go
/*
 * Name: Abdelrahman Ahmed
 * Professor: Dr. Dai
 * Course: CS416 - Software Engineering
 * Date: Feb 17, 2026
 *
 * File: MergeSort.go
 * Description:
 *   Implements Merge Sort using a divide-and-conquer strategy that recursively splits and merges sorted halves.
 *   It runs in O(n log n) time and is stable, but uses extra memory for merging.
 */

package main

import "fmt"

// MergeSort returns a NEW sorted slice in ascending order (does not modify input).
func MergeSort(a []int) []int {
	if len(a) <= 1 {
		out := make([]int, len(a))
		copy(out, a)
		return out
	}
	mid := len(a) / 2
	left := MergeSort(a[:mid])
	right := MergeSort(a[mid:])
	return merge(left, right)
}

func merge(left, right []int) []int {
	out := make([]int, 0, len(left)+len(right))
	i, j := 0, 0
	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			out = append(out, left[i])
			i++
		} else {
			out = append(out, right[j])
			j++
		}
	}
	out = append(out, left[i:]...)
	out = append(out, right[j:]...)
	return out
}
func main() {
	test := []int{9, 3, 5, 1, 8, 2, 7, 4, 6}
	fmt.Println("Before:", test)
	sorted := MergeSort(test)
	fmt.Println("After: ", sorted)
	fmt.Println("Original unchanged:", test)
}
